### PR TITLE
Fix RefitLevel using last level as the base level even when these two can be different

### PR DESCRIPTION
**Pull request by hx235**
Context/Summary: Credit to @cbi42 for finding the bug in https://github.com/facebook/rocksdb/pull/10988/files#r1048004670. Basically as titled.

Test:
Make check

### CODE REVIEW

File: db/db_impl/db_impl_compaction_flush.cc [hunk 1]
```diff
@@ -1177,7 +1177,9 @@ Status DBImpl::CompactRangeInternal(const CompactRangeOptions& options,
           break;
         }
         if (output_level == ColumnFamilyData::kCompactToBaseLevel) {
-          final_output_level = cfd->NumberLevels() - 1;
+          assert(cfd->current() && cfd->current()->storage_info());
+          VersionStorageInfo* vstorage = cfd->current()->storage_info();
```
Comment (cbi42) [inline db/db_impl/db_impl_compaction_flush.cc:1181]: Do we need mutex_ or referenced super version for this? I'm not sure if this is thread safe.

File: HISTORY.md [hunk 2]
```diff
@@ -13,6 +13,7 @@
 * Fixed a BackupEngine bug in which RestoreDBFromLatestBackup would fail if the latest backup was deleted and there is another valid backup available.
 * Fix L0 file misorder corruption caused by ingesting files of overlapping seqnos with memtable entries' through introducing `epoch_number`. Before the fix, `force_consistency_checks=true` may catch the corruption before it's exposed to readers, in which case writes returning `Status::Corruption` would be expected. Also replace the previous incomplete fix (#5958) to the same corruption with this new and more complete fix.
 * Fixed a bug in LockWAL() leading to re-locking mutex (#11020).
+* Fixed a bug that `RefitLevel()` can move files from wrong level when base level is different from the last level under universal compaction and `Options::level_compaction_dynamic_level_bytes > 0`.
```
Comment (ajkr) [inline HISTORY.md:16]: "universal compaction and `Options::level_compaction_dynamic_level_bytes > 0`" looks like a contradiction. I didn't see universal compaction mentioned in the description, title, or linked discussion.

File: db/db_impl/db_impl_compaction_flush.cc (hunk 1 above)
Comment (cbi42) [reply db/db_impl/db_impl_compaction_flush.cc:1182]: One other concern I realize now is if some other compaction that changed the based level that ran after `RunManualCompaction()` but before we get base_level here.

File: db/db_impl/db_impl_compaction_flush.cc [hunk 3]
```diff
@@ -1177,7 +1177,11 @@ Status DBImpl::CompactRangeInternal(const CompactRangeOptions& options,
           break;
         }
         if (output_level == ColumnFamilyData::kCompactToBaseLevel) {
-          final_output_level = cfd->NumberLevels() - 1;
```
Comment (ajkr) [inline db/db_impl/db_impl_compaction_flush.cc:1180]: The old code assumed the final output level of a `CompactRange()` with `level_compaction_dynamic_level_bytes==true` would be the last level. That assumption seems reasonable because dynamic leveled compaction is supposed to fill in levels starting with the last level, and `CompactRange()` is supposed to compact down to the last level containing any data (or last level if all of L1+ are empty). 

However as you found there can be edge cases where the final output level is not the last level. In the unit test it looks like the cause was a DB that already had data in L1+ reopened with `num_levels_new > num_levels_old` (`num_levels_old` is the pre-reopen value of `num_levels`, and `num_levels_new` is the post-reopen value). Then the target level of the post-reopen `CompactRange()` is `num_levels_old-1`, which is not the last level anymore. Another scenario it happens, often surprising/inconveniencing users, is after enabling `level_compaction_dynamic_level_bytes`. That makes the migration to enable `level_compaction_dynamic_level_bytes` difficult as even compacting the whole DB through `CompactRange()` with nullptr endpoints is not enough.

In my opinion, `CompactRange()` should move the data to the ideal level according to the current config, i.e., not stop at empty level when `level_compaction_dynamic_level_bytes==true`. I think the type of change in this PR adds complexity and has race conditions (base level can change after `RunManualCompaction()`). If there's a way to do it without the race conditions it might be worthwhile as a stop-gap solution. The conceptual simplicity of the old code will be missed (by me), though.

**Review by ajkr — CHANGES_REQUESTED** [review]

File: HISTORY.md (hunk 2 above)
Comment (hx235) [reply HISTORY.md:16]: universal compaction was a typo from me, supposed to be level compaction

**Comment by hx235** [comment]
Since we have found out more complexity than I originally anticipated as a quick fix, I will postpone fixing this. Will reopen when I have time and idea on how to fix this.

**Comment by ajkr** [comment]
If we do T117522570 to ensure last level is nonempty during open when there is data in L1+, and possibly another change to make it stay nonempty when there is data in L1+, then the current code would become correct. I think we'd have less LSM shape operational challenges with that approach. OTOH, there's not much time to do that, so quick fixes are nice too if they happen to be simple/correct

File: db/db_impl/db_impl_compaction_flush.cc (hunk 3 above)
Comment (cbi42) [reply db/db_impl/db_impl_compaction_flush.cc:1180]: > CompactRange() is supposed to compact down to the last level containing any data

I think the current behavior is to compact down to the lowest level which contains any data *that overlaps with the key range passed into `CompactRange()`*. So there are more cases where the old code would move wrong the wrong level during refit level.
